File: backend/app/api/endpoints/finance.py
import uuid
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.api.dependencies import get_db, get_current_user
from app.core.db.models import User
from app.schemas.timesheet import TimesheetOut
from app.schemas.invoice import FinanceApprove, FinanceReject, InvoiceOut
from app.services import finance_service
from app.services.invoice_engine import InvoiceConfig

router = APIRouter()


@router.get("/pending", response_model=list[TimesheetOut])
def get_finance_pending(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Finance queue: all client_approved timesheets awaiting Finance review."""
    return finance_service.get_finance_pending_timesheets(db)


@router.post("/{timesheet_id}/approve", response_model=InvoiceOut, status_code=status.HTTP_201_CREATED)
def finance_approve(
    timesheet_id: uuid.UUID,
    payload: FinanceApprove,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Finance approves a timesheet.
    Locks the timesheet and triggers the Invoice Engine to generate an invoice.
    Returns the generated invoice.
    """
    config = InvoiceConfig(
        hourly_rate=payload.hourly_rate,
        tax_rate=payload.tax_rate,
        payment_terms=payload.payment_terms,
        notes=payload.notes,
    )
    return finance_service.finance_approve_timesheet(db, timesheet_id, current_user.id, config)


@router.post("/{timesheet_id}/reject", response_model=TimesheetOut)
def finance_reject(
    timesheet_id: uuid.UUID,
    payload: FinanceReject,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Finance rejects a timesheet with a required reason. Returns the timesheet to the Candidate."""
    return finance_service.finance_reject_timesheet(db, timesheet_id, current_user.id, payload.reason)

File: frontend/src/lib/finance.ts
import { api } from './api';
import { Timesheet } from './timesheets';

export interface FinanceApprovePayload {
  hourly_rate?: number;
  tax_rate?: number;
  payment_terms?: string;
  notes?: string;
  client_name: string;
  billing_contact?: string;
  billing_address?: string;
  billing_email?: string;
}

export interface FinanceRejectPayload {
  reason: string;
}

export interface DashboardStats {
  pending_validation: number;
  draft_invoices: number;
  ready_invoices: number;
  sent_invoices: number;
  paid_invoices: number;
  total_revenue: number;
  total_outstanding: number;
}

export interface FinanceTrendDataPoint {
  date: string;
  draft: number;
  sent: number;
  paid: number;
}

export const financeApi = {
  /** Finance queue: all client_approved timesheets */
  getPending: async (): Promise<Timesheet[]> => {
    return api.get<Timesheet[]>('/finance/pending');
  },

  /** Get a specific timesheet for finance review */
  getTimesheetById: async (timesheetId: string): Promise<Timesheet> => {
    return api.get<Timesheet>(`/finance/timesheets/${timesheetId}`);
  },

  /** Finance approves — triggers invoice generation. Returns generated Invoice. */
  approve: async (timesheetId: string, payload: FinanceApprovePayload): Promise<Invoice> => {
    return api.post<Invoice, FinanceApprovePayload>(`/finance/${timesheetId}/approve`, payload);
  },

  /** Finance rejects — returns timesheet to candidate */
  reject: async (timesheetId: string, reason: string): Promise<Timesheet> => {
    return api.post<Timesheet, FinanceRejectPayload>(`/finance/${timesheetId}/reject`, { reason });
  },

  /** Fetch dashboard stats */
  getDashboardStats: async (): Promise<DashboardStats> => {
    return api.get<DashboardStats>('/finance/dashboard/stats');
  },

  /** Fetch trend data for invoices */
  getTrendData: async (): Promise<{ data: FinanceTrendDataPoint[] }> => {
    return api.get<{ data: FinanceTrendDataPoint[] }>('/finance/dashboard/trend');
  },
};

export interface Invoice {
  id: string;
  invoice_number: string;
  status: 'draft' | 'ready' | 'sent' | 'payment_pending' | 'paid';
  currency?: string;
  template_name?: string;
  snapshot_data?: any;
  timesheet_id: string;
  candidate_id: string;
  generated_by_id: string;
  period_start_date: string;
  period_end_date: string;
  total_hours: number;
  hourly_rate: number;
  subtotal: number;
  tax_rate: number;
  tax_amount: number;
  total_amount: number;
  work_summary: string;
  payment_terms: string;
  notes: string | null;
  issued_at: string;
  due_date: string;
  created_at: string;
  updated_at: string;
}
